Add folder navigation to files / image selector interfaces

paescuj opened the pull request:
Closes #15968 / ENG-743

<table>
<tr>
<td align="center"><b>Before</b> (All files)</td>
<td align="center"><b>After</b> (Browsable by folders)</td>
</tr>
<tr>
<td><img  src="https://user-images.githubusercontent.com/5363448/234709260-ff30d23b-5e1a-4029-ac55-bac9f6755031.png"></td>
<td><img src="https://user-images.githubusercontent.com/5363448/234709248-308cec7c-0e74-4488-91a0-51de321182e8.png"></td>
</tr>
</table>

https://user-images.githubusercontent.com/5363448/234712523-d086961b-8794-4aa4-99d5-afa6cd8b9bab.mp4

paescuj committed:
Add folder picker to file / image drawer

paescuj committed:
Update v-upload for image fields

paescuj committed:
Fix filter for multiple files interface

paescuj committed:
Separate state for open folders in drawers

paescuj committed:
Lint fixes

paescuj committed:
Merge branch 'main' into folder-picker-files-drawer

paescuj committed:
Mobile header & disable folder actions in files drawer

paescuj committed:
Fix filter in drawer for files interface

paescuj committed:
Adjust styles instead of logic to ensure filter has enough space

paescuj committed:
Pass folder actionsDisabled option to child folders

paescuj committed:
Rmoeve unncessary style prop in card actions

paescuj committed:
Disable folder picker when default folder selected

To follow the same behavior as before and not introducting any breaking change in this regard.
This can still be enhanced later on, e.g. with new options "enable
folder picker" and "show only subfolders"

paescuj committed:
Make the sidebar scrollable (when many folders)

jaads reviewed (approved):
Looks very good to me and works as expected 🎉

rijkvanzanten committed:
Merge branch 'main' into folder-picker-files-drawer

paescuj committed:
Merge branch 'main' into folder-picker-files-drawer

paescuj committed:
Create stale-plants-jam.md

danieljwestman commented:
Very much needed improvent! Good work 💪🏻

paescuj committed:
Add title prop for elements which might be clipped

paescuj committed:
Fix nav for special routes (item, not found) & when drawer not unmounted

paescuj committed:
Revert header style changes, find other solution...

paescuj committed:
Apply custom filter to system filter to fix item count for files intf

paescuj committed:
Consolidate 'Special' type

azrikahar commented on app/src/views/private/components/drawer-files.vue:
```diff
@@ -0,0 +1,257 @@
+	</component>
+</template>
+
+<script lang="ts">
```
With #18293 in mind, let's use script setup for new components going forward as well 👍

paescuj committed:
Update stale-plants-jam.md

azrikahar commented on app/src/views/private/components/files-navigation.vue:
```diff
@@ -0,0 +1,172 @@
+	if (props.customTargetHandler) {
+		props.customTargetHandler(target);
+	} else {
+		const path = ['files'];
```
super tiny thing, but would `paths` be more apt here to slightly indicate it's an array?

LasseRosenow commented:
What happens if we have a selected folder that itself also contains folders?

azrikahar commented on app/src/views/private/components/header-bar.vue:
```diff
@@ -16,7 +16,7 @@
 				<slot name="title">
 					<slot name="title:prepend" />
-					<h1 class="type-title">{{ title }}</h1>
+					<h1 class="type-title" :title="title">{{ title }}</h1>
```
Presumably this is to resolve text that are "cut off" due to not enough space, coincidentally I've been wondering about this recently! 

I think we should probably use the `<v-text-overflow :text="title" placement="bottom">` from a design consistency perspective as we use that throughout the app. (understandably I do also assume the title attribute would have technically be better in terms of a11y if I'm not mistaken...)

paescuj replied:
Yesss 🚀

paescuj commented:
> What happens if we specified a concrete folder in the file interface that itself also contains folders?

Only the files inside the selected folder are shown. This keeps up with the previous behavior, mainly to not introduce any breaking change in this regard. 
We will probably make this configurable in a separate step, because we still have to think about how we want to implement this, but I can imagine that the following options will be offered then:
- No navigation
- Limited navigation (only show sub folders)
- Full navigation (show all folders)

paescuj commented on app/src/components/v-nav.vue:
Oh, good one, wasn't aware of this! Will take a peek at it 👍

paescuj committed:
Update app/src/components/v-nav.vue

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

paescuj commented on app/src/utils/get-folder-filter.ts:
```diff
@@ -0,0 +1,48 @@
+import { Filter } from '@directus/types';
+import { subDays } from 'date-fns';
+
+export function getFolderFilter(special?: string, folder?: string, currentUserId?: string) {
```
I agree, we certainly profit more from type safety this way 🦺

paescuj committed:
Explicit return type for getFolderFilter util

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

paescuj committed:
Use v-text-overflow components for title

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

azrikahar commented:
Seems like there's a visual bug where an expanded search input tries to "push" to the right instead of the left, thus causing the actions buttons to be out of the view:

https://user-images.githubusercontent.com/42867097/234909120-92534478-53ff-4764-951d-eed337da9df4.mp4

paescuj commented:
> Seems like there's a visual bug where an expanded search input tries to "push" to the right instead of the left, thus causing the actions buttons to be out of the view:
> chrome_LsQEabm1u2.mp4

Yep, I'm still looking at how I can best solve this... 👍

rijkvanzanten commented:
Container queries? Container queries!!

paescuj commented:
> Container queries? Container queries!!

Yeeeeahh, wanting this soo much!

rijkvanzanten commented:
> > Container queries? Container queries!!
> 
> Yeeeeahh, wanting this soo much!

<img width="602" alt="CleanShot 2023-04-27 at 13 05 42@2x" src="https://user-images.githubusercontent.com/9141017/234937130-b3b347d2-7947-4139-80e3-c5006b1c02d3.png">

Lets goooooooooooooooooooooooooooooooooo 

![](https://media.giphy.com/media/13G7hmmFr9yuxG/giphy.gif)

licitdev reviewed (commented):
Noticed something small which may not be an issue. There is a "caching" behaviour for the `image` and `files` interfaces. For "image", it clears once an image is selected. However for "files", the caching remains after a selection.

https://user-images.githubusercontent.com/26413686/234940824-019fcf79-494e-4006-bfb7-6f0d21d8000e.mov

paescuj commented:
> Noticed something small which may not be an issue. There is a "caching" behaviour for the `image` and `files` interfaces. For "image", it clears once an image is selected. However for "files", the caching remains after a selection.
> Screen.Recording.2023-04-28.at.1.21.03.AM.mov

This is because in case of the `image` and `files` interfaces the drawer isn't unmounted. In the same way, the filter value is retained for these interfaces, which is also the case for the M2O interface. However, I would not consider it an issue. If we want to change this behavior, I would approach it in a separate pull request and then we would arguably have to adjust it for the M2O as well.

br41nslug commented on app/src/types/folders.ts:
```diff
@@ -0,0 +1 @@
+export type Special = 'all' | 'mine' | 'recent';
```
What does `Special` mean in this context? Could we name this something more specific to prevent future name collision?
Feels like it could be `FolderFilterType` or something similar 🤔

paescuj replied:
It's for these special folder views:
<img width="209" alt="Screenshot 2023-05-01 at 15 18 43" src="https://user-images.githubusercontent.com/5363448/235457095-94972de7-e79a-488a-af4a-a3634564d072.png">

Probably `SpecialFolder`, `SpecialFolderView`? 🤔

br41nslug replied:
Both are more descriptive than just `Special` :)

br41nslug commented on app/src/views/private/components/header-bar-actions.vue:
```diff
@@ -84,7 +84,7 @@ export default defineComponent({
 @media (min-width: 960px) {
 	.actions .action-buttons .sidebar-toggle {
-		display: none;
+		display: none !important;
```
This `!important` does not look to be required on my end (nor does this entire `.side-bar-toggle` block for that matter 😬 ). is this a workaround for a specific browser?
![image](https://user-images.githubusercontent.com/9389634/235453760-b800e6e5-04b0-40da-a3d7-167741e1a98d.png)

paescuj replied:
Good eyes! It's a little unrelated fix, noticed while working on this feature. Easier to explain with the following videos 😄 

**Before**

https://user-images.githubusercontent.com/5363448/235456658-33a3121b-9177-460a-be92-f357c29337f4.mp4

**After**

https://user-images.githubusercontent.com/5363448/235456662-5cf5836c-0624-46ff-a98c-ba121bbc44ed.mp4

br41nslug reviewed (commented):
A very much wanted feature :) looking good but i do have some feedback in the comments below.

Also out of the scope of this PR but i ran into a couple of point of improvement for later (these apply to both the drawer and the files module):
- Adding a loading state when switching folders because it can take a bit for them to render

https://user-images.githubusercontent.com/9389634/235454167-40e4efd7-009e-460b-8d44-d47ef7671926.mp4

- A solution like less indentation for deeply nested folders, the sidebars are resizable but only up to a point

![firefox_eZxnazzCpa](https://user-images.githubusercontent.com/9389634/235454273-ca61a8fd-ab33-483e-8849-3230eed395db.png)

- An option to select subfolders of the selected root folder in the files interfaces, currently we can select a folder from which you can select files but you cannot see the files nested in subfolders

paescuj committed:
Don't pass initial selected items as they are filtered out anyway

Led to a bad UX, as the user was shown that items were selected (e.g. "2 selected items") but these were neither visible nor could be deselected.
This issue has existed before.

paescuj committed:
Cleanup

- Extend drawer-files from drawer-collection to prevent code duplication
- Reuse and clarify types (SpecialFolder, FolderTarget)
- Simplify logic for resetOpenFolders (now localOpenFolders)
- Migrate new / relocated components to 'script setup'

paescuj committed:
Merge branch 'main' into folder-picker-files-drawer

paescuj committed:
Update changeset

paescuj committed:
Update stale-plants-jam.md

rijkvanzanten committed:
Merge branch 'main' into folder-picker-files-drawer

br41nslug commented on app/src/views/private/components/drawer-collection.vue:
```diff
@@ -16,15 +18,20 @@
 			<template #title-outer:prepend>
 				<v-button class="header-icon" rounded icon secondary disabled>
-					<v-icon :name="collectionInfo.icon" :color="collectionInfo.color" />
+					<v-icon :name="collectionInfo?.icon" :color="collectionInfo?.color" />
```
```js
<v-icon :name="collectionInfo?.icon" :color="collectionInfo?.color" />
```
if `collectionInfo` is undefined here should we not hide the icon component altogether? 🤔

rijkvanzanten replied:
If collectionInfo is undefined, this whole page shouldn't render (it should be a 404) 🤔

br41nslug replied:
Should the type not reflect that then 🤔 currently it can be null all over this component according to the typing

rijkvanzanten replied:
Yup! Welcome to the TypeScript Modernization Program for the Application, we're happy to have you. Keep in mind that Vue didn't really support TypeScript back when we started it, so a lot of these types have been added in over time. This is one of many rough edges that we'll be cleaning up over time, starting with the script[setup] business 🙂

rijkvanzanten commented:
# Todos

- [ ] Solve merge conflicts
- [ ] Fix style quirks due to layouts rendering in a smaller container (container queries)

br41nslug committed:
Merge branch 'main' into folder-picker-files-drawer

br41nslug committed:
re-added missing navigation files

br41nslug committed:
fix linter issues

br41nslug committed:
Merge branch 'main' into folder-picker-files-drawer

br41nslug committed:
fixing bad navigation merge

br41nslug committed:
Merge branch 'folder-picker-files-drawer' of https://github.com/directus/directus into folder-picker-files-drawer

licitdev committed:
Fix missing navigation-folder component

licitdev commented on app/src/views/private/components/files-navigation.vue:
```diff
@@ -62,32 +65,54 @@
-import NavigationFolder from './navigation-folder.vue';
+import { ref, watch } from 'vue';
+import { useFolders, openFoldersInitial } from '@/composables/use-folders';
+import NavigationFolder from './files-navigation-folder.vue';
```
```suggestion
import NavigationFolder from '@/views/private/components/files-navigation-folder.vue';
```

Should this be updated for consistency?

paescuj committed:
Prevent sidebar from being resizable for now

paescuj committed:
Make it a minor bump for "directus"
